### PR TITLE
Fix null via not pushdown

### PR DESCRIPTION
**Pull request by nicklan**
`NOTNULL` is tricky to compute as a data skipping filter. It's NOT the same as just inverting the `NULL` check. 

For example, to check if a col could be null, we can see if `nullCount > 0` (assuming tight bounds), but `NOTNULL` is NOT the same as `nullCount <= 0`, for that we need `nullCount < numRecords` (i.e. at least one record is not null)

The wide bounds expressions _do_ invert, but this keeps them as separate for clarity, and because it opens the possibility of not needing `NOT` in the expression evaluator.

Additionally inverting gt/lt style expressions also isn't correct. This fixes that too.

So let's just do what we suggested in #246 and push down, so we can have two different checks.

Also adds lots more tests for various combinations of predicates.

### CODE REVIEW

**Comment by nicklan** [comment]
@scovich fyi, #231 was actually broken for partitions that had a mix of null and not null :( this fixes that

**Comment by scovich** [comment]
> https://github.com/delta-incubator/delta-kernel-rs/pull/231 was actually broken for partitions that had a mix of null and not null

Yikes, I totally forgot about that reason for pushing NOT down! It's not just IS NULL, either -- inverting a data skipping expression is _usually_ not logically equivalent to creating a data skipping expression from an inverted input. For example, data skipping expression for `x < 10` is `add.stats.minValues < 10`, but the data skipping expresion for `NOT(x < 10)` = `x >= 10` is `add.stats.maxValues >= 10`.

We probably need to add unit tests for all these cases, not just the IS [NOT] NULL case.

File: kernel/src/scan/data_skipping.rs [hunk 1]
```diff
@@ -44,22 +44,76 @@ fn get_wide_null_expr(null_col: String) -> Expr {
...
-///   operands that are not eligible for data skipping.
-/// - `OR` is rewritten only if all operands are eligible for data skipping. Otherwise,
-///   the whole OR expression is dropped.
+/// - `AND` is rewritten as a conjunction of the rewritten operands where we just skip operands that
+///         are not eligible for data skipping.
+/// - `OR` is rewritten only if all operands are eligible for data skipping. Otherwise, the whole OR
+///        expression is dropped.
 fn as_data_skipping_predicate(expr: &Expr) -> Option<Expr> {
     use BinaryOperator::*;
     use Expr::*;
```
Comment (scovich) [inline kernel/src/scan/data_skipping.rs:119]: aside: Looking at the `Equal` and `NotEqual` cases below, I'm a bit surprised they're not consistent with each other? `x = 10` converts to `x <= 10 AND 10 <= x` (which we then extract a data skipping expression for), while `x != 10` converts directly to a data skipping expression? It seems like the former is a simpler/safer approach?

(also, they could have used `Expr::and` and `Expr::or` instead of `and_from` and `or_from`).

File: kernel/src/scan/data_skipping.rs [hunk 2]
```diff
@@ -97,8 +151,8 @@ fn as_data_skipping_predicate(expr: &Expr) -> Option<Expr> {
             op: UnaryOperator::Not,
```
Comment (scovich) [inline kernel/src/scan/data_skipping.rs:151]: qq: if we `use UnaryOperator::*` would that allow `UnaryOperation { op: Not, expr } =>{` on one line? 
Or does rustfmt still force it to multiple lines just because we constrained the type?

**Review by scovich — COMMENTED** [review]
The change looks good, but I strongly suspect unit test coverage is lacking. Should we add the missing tests now or defer it as a follow-up?

**Comment by nicklan** [comment]
> > #231 was actually broken for partitions that had a mix of null and not null
> 
> Yikes, I totally forgot about that reason for pushing NOT down! It's not just IS NULL, either -- inverting a data skipping expression is _usually_ not logically equivalent to creating a data skipping expression from an inverted input. For example, data skipping expression for `x < 10` is `add.stats.minValues < 10`, but the data skipping expression for `NOT(x < 10)` = `x >= 10` is `add.stats.maxValues >= 10`.
> 
> I believe this PR fixes everything, but we probably need to add unit tests for all these cases, not just the IS [NOT] NULL case.

Yeah, it was really broken before :)

More tests are totally warranted. I've added a bunch for single operators, but should still add some for AND/OR and inversions of that. I'll get those into this PR in the next few days so we can (hopefully) just be sure that things work.

File: kernel/src/expressions/mod.rs [hunk 3]
```diff
@@ -36,12 +36,39 @@ pub enum BinaryOperator {
     Distinct,
 }
 
+impl BinaryOperator {
+    // invert an operator. Returns Some<InvertedOp> if the operator supports inversion, None if it
+    // cannot be inverted
+    pub(crate) fn invert(&self) -> Option<BinaryOperator> {
```
[the remark this answers is not in the document]
Comment (nicklan) [reply kernel/src/expressions/mod.rs:42]: Are you just thinking that some of the data_skipping expr construction could be simplified with `commute`? We do need to commute there, but we also change the actual expressions (i.e. the column name becomes `stats.maxValues.colName` or similar) so since we have to rebuild anyway I'm not sure `commute` is needed?

File: kernel/src/scan/data_skipping.rs (hunk 1 above)
Comment (nicklan) [reply kernel/src/scan/data_skipping.rs:119]: Yeah, let me see about simplifying this

**Comment by scovich** [comment]
> More tests are totally warranted. I've added a bunch for single operators, but should still add some for AND/OR and inversions of that. I'll get those into this PR in the next few days so we can (hopefully) just be sure that things work.

You may want to take a peek at delta-spark [DataSkippingDeltaTests.scala](https://github.com/delta-io/delta/blob/master/spark/src/test/scala/org/apache/spark/sql/delta/stats/DataSkippingDeltaTests.scala) -- it has a whole bunch of test cases that might be useful to replicate. Eventually we should probably harmonize them into DAT?

File: kernel/tests/read.rs [hunk 4]
```diff
@@ -503,23 +516,111 @@ fn column_ordering_and_projection() -> Result<(), Box<dyn std::error::Error>> {
...
+    }
+    res.push("+---------+--------+".to_string());
+    res
+}
+
+#[test]
+fn predicate_on_number() -> Result<(), Box<dyn std::error::Error>> {
+    let cases = vec![
+        (
+            Expression::Column("number".to_string()).lt(Expression::Literal(Scalar::Long(4))),
```
Comment (scovich) [inline kernel/tests/read.rs:540]:
```suggestion
            Expression::column("number").lt(Expression::literal(4i64)),
```
Code volume can be reduced quite a bit this way... might even be worth a `use Expression::{column, literal}` to allow
```suggestion
            column("number").lt(literal(4i64)),
```

File: kernel/tests/read.rs [hunk 5]
```diff
@@ -503,23 +516,111 @@ fn column_ordering_and_projection() -> Result<(), Box<dyn std::error::Error>> {
...
+    Ok(())
+}
+
+#[test]
+fn predicate_on_number_not() -> Result<(), Box<dyn std::error::Error>> {
+    let cases = vec![
+        (
+            Expression::Column("number".to_string())
+                .lt(Expression::Literal(Scalar::Long(4)))
+                .not(),
```
Comment (scovich) [inline kernel/tests/read.rs:582]: I find this form of `not` rather non-intuitive. Perhaps better to just do functional form instead?
```suggestion
            not(lt(column("number"), literal(4i64)),
```

File: kernel/tests/read.rs (hunk 4 above)
Comment (nicklan) [reply kernel/tests/read.rs:540]: You can't `use` struct methods like that, but yeah, just using the function forms is much better

**Comment by nicklan** [comment]
> You may want to take a peek at delta-spark [DataSkippingDeltaTests.scala](https://github.com/delta-io/delta/blob/master/spark/src/test/scala/org/apache/spark/sql/delta/stats/DataSkippingDeltaTests.scala) -- it has a whole bunch of test cases that might be useful to replicate. Eventually we should probably harmonize them into DAT?

Yep, created #249 as follow-up work

File: kernel/src/scan/data_skipping.rs (hunk 1 above)
Comment (nicklan) [reply kernel/src/scan/data_skipping.rs:119]: This was acting odd, and I don't have time to fix it right now. I'd like to merge asap since things are "broken", so I've created #250 to remind us to look at this

File: kernel/src/expressions/mod.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (nicklan) [reply kernel/src/expressions/mod.rs:55]: Done

File: kernel/tests/read.rs [hunk 6]
```diff
@@ -706,8 +694,143 @@ fn mixed_not_null() -> Result<(), Box<dyn std::error::Error>> {
...
     Ok(())
 }
+
+#[test]
+fn and_or_predicates() -> Result<(), Box<dyn std::error::Error>> {
+    let cases = vec![
+        (
+            Expression::column("number")
+                .gt(Expression::literal(4i64))
+                .and(Expression::column("a_float").gt(Expression::literal(5.5))),
```
Comment (scovich) [inline kernel/tests/read.rs:709]: nit: AND/OR are similar to NOT, in being hard(er) to read in this "chained" form.
```suggestion
            Expression::and(
                Expression::column("number").gt(Expression::literal(4i64)),
                Expression::column("a_float").gt(Expression::literal(5.5)),
            ),
```
It does admittedly add one extra line to the expression tho. Your call which is better overall.

Comment (scovich) [reply kernel/tests/read.rs:709]: Also, just double checking: `a_float` is a f64 column (not f32), right?

**Review by scovich — APPROVED** [review]
Approving to un-break. We can quibble about readability tweaks and tests as a follow-up.

**Review by hntd187 — APPROVED** [review]
LGTM, nothing Ryan didn't already mention, I did write some small tests for my own sanity, but they came out fine
